### ISSUE TITLE
Could not look up qualified variable '$::consul_version'

### ISSUE THREAD

**Issue by HanSooloo** [solarkennedy/puppet-consul#327]
Using r10k pattern to install a consul agent and seeing the above as a warning during install.  Wondering what I'm doing wrong, or whether this is safe to ignore.

Is there a hierarchy/namespace error in the way I call on the `::consul` class from within the role definition?

Relevant log snip:
```
==> bootstrap0: Warning: Scope(Class[Consul::Install]): Could not look up qualified variable '$::consul_version'; class $ could not be found
==> bootstrap0: Notice: Compiled catalog for bootstrap.role.com in environment dev in 2.82 seconds
==> bootstrap0: Notice: /Stage[main]/Consul::Install/Group[consul]/ensure: created
==> bootstrap0: Notice: /Stage[main]/Consul::Install/User[consul]/ensure: created
==> bootstrap0: Notice: /Stage[main]/Consul::Install/File[/opt/consul]/ensure: created
==> bootstrap0: Notice: /Stage[main]/Consul::Install/File[/opt/consul/archives]/ensure: created
==> bootstrap0: Notice: /Stage[main]/Consul::Install/File[/opt/consul/archives/consul-0.7.0]/ensure: created
==> bootstrap0: Notice: /Stage[main]/Consul::Install/Archive[/opt/consul/archives/consul-0.7.0.zip]/ensure: download archive from https://releases.hashicorp.com/consul/0.7.0/consul_0.7.0_linux_amd64.zip to /opt/consul/archives/consul-0.7.0.zip and extracted in /opt/consul/archives/consul-0.7.0 with cleanup
==> bootstrap0: Notice: /Stage[main]/Consul::Install/File[/opt/consul/archives/consul-0.7.0/consul]/mode: mode changed '0755' to '0555'

```

```
**project_root**
+-puppet
    +-hiera
    |   +-bootstrap.role.com.yaml
    +-manifests
    |   +-site.pp
    +-site
    |   +-profiles
    |   |   +-manifests
    |   |       +- consul
    |   |            +-agent.pp
    |   +-roles
    |       +-manifests
    |           +- bootstrap.pp
    +-environment.conf
    +-Puppetfile
```

```
**hiera/bootstrap.role.com.yaml**
---
environment: production
classes:
  - roles::bootstrap

```

```
**manifests/site.pp**
hiera_include('classes')

```

```
**site/profiles/manifests/consul/agent.pp**
# Configure consul
class profiles::consul::agent {
  class { '::consul':
    config_hash => {
      'datacenter'  => 'dc1',
      'data_dir'    => '/opt/consul',
      'client_addr' => '0.0.0.0',
      'log_level'   => 'INFO',
      'node_name'   => $::hostname,
      'bind_addr'   => $::ipaddress_eth1,
      'server'      => false,
      'retry_join'  => [hiera('join_addr')],
    }
  }
}

```

```
**roles/manifests/bootstrap.pp**
# Bootstrap server role
class roles::bootstrap {
    include profiles::consul::agent
}

```

```
**environment.conf**
manifest = manifests/site.pp
modulepath = modules:site

```

```
**Puppetfile**
forge "https://forge.puppetlabs.com"

# common
mod 'puppetlabs/stdlib'
mod 'puppet/archive'
mod 'nanliu/staging'

# consul
mod 'KyleAnderson/consul'

```

**Comment by solarkennedy** [comment]
It's "sorta" safe to ignore. You can't access the fact early if consul isn't installed yet.

You could use `getvar()` to lookup variables that may not exist:
https://forge.puppet.com/puppetlabs/stdlib#getvar

**Comment by HanSooloo** [comment]
Thank you for the quick response!

Kind of like a "can't tell you the currently installed version before it is installed" situation then?

How could I incorporate the getvar call into my code?

**Comment by solarkennedy** [comment]
Oh, this is in our code, not yours.
This is fixed 
https://github.com/solarkennedy/puppet-consul/commit/7d614aea677f701cfe5913fda3c6036d29774680

Try using the latest version?

**Comment by HanSooloo** [comment]
I am apparently using release 2.0.1 as pulled by the Puppetfile.  Wondering if there is some silly hierarchy/namespace issue (have to admin, new to the world of Puppet!)

**Comment by solarkennedy** [comment]
It is good practice to pin the modules to ensure reproducible results. Please pin to `2.1.0`. (are you sure you are on `2.0.1`? I've never released such a version)

**Comment by HanSooloo** [comment]
The plot thickens .... the TGZ listed here (https://forge.puppet.com/v3/files/KyleAnderson-consul-2.1.0.tar.gz) contains a `CHANGELOG.md` that references a `2.0.1` release: https://forge.puppet.com/KyleAnderson/consul

Probably a typo in the text file, but weird that the TGZ has it, but GitHub has no recollection of a `2.0.1` text literal in any commit.

![image](https://cloud.githubusercontent.com/assets/879582/23957797/9c4a88f0-0976-11e7-81f5-b7b3d379e911.png)

**Back to the real question:** I was actually able to get rid of the warning message by manually applying https://github.com/solarkennedy/puppet-consul/commit/f7466a5faf18c150639d9ea0b17755db96c64872 (guess it just removes the `$` sign for root level reference?)

Also, https://github.com/solarkennedy/puppet-consul/archive/v2.1.0.tar.gz on GitHub has a `CHANGELOG.md` that references `2.0.0`
![image](https://cloud.githubusercontent.com/assets/879582/23958747/74d75296-0979-11e7-98a4-ff18bd4fbf6c.png)

**Comment by solarkennedy** [comment]
Hmm. When I make the changelog files it is still "manual" (run the script) so it is easy for me to forget and update them later. Forget the changelog. If a release is tagged as 2.1.0, it *is* 2.1.0.

Anyway, I've released 2.1.1 (and I updated the changelog first) and now it has that f74 commit in it. Pin to 2.1.1 and that should be it.

**Comment by HanSooloo** [comment]
Works perfectly without any warning messages with 2.1.1.  Thanks for the quick response 👍
